# Analyzing German eBay Car Sales Data

In this project, we will be working on a dataset of used cars from _eBay Kleinanzeigen_, a [classifieds](https://en.wikipedia.org/wiki/Classified_advertising) section of the German eBay website.

The dataset was originally scraped and uploaded to [Kaggle](https://www.kaggle.com/). The version of the dataset we are working with is a sample of 50,000 data points that was prepared by [Dataquest](https://www.dataquest.io/), and with more dirty data. The original dataset, can be found [here](https://data.world/data-society/used-cars-data).

The data dictionary is as follows:

- `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
- `name` - Name of the car.
- `seller` - Whether the seller is private or a dealer.
- `offerType` - The type of listing
- `price` - The price on the ad to sell the car.
- `abtest` - Whether the listing is included in an A/B test.
- `vehicleType` - The vehicle Type.
- `yearOfRegistration` - The year in which the car was first registered.
- `gearbox` - The transmission type.
- `powerPS` - The power of the car in PS.
- `model` - The car model name.
- `kilometer` - How many kilometers the car has driven.
- `monthOfRegistration` - The month in which the car was first registered.
- `fuelType` - What type of fuel the car uses.
- `brand` - The brand of the car.
- `notRepairedDamage` - If the car has a damage which is not yet repaired.
- `dateCreated` - The date on which the eBay listing was created.
- `nrOfPictures` - The number of pictures in the ad.
- `postalCode` - The postal code for the location of the vehicle.
- `lastSeenOnline` - When the crawler saw this ad last online.

The aim of this project is to clean the data and analyze the included used car listings. 

## Reading in the Data File


In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv('autos.csv', encoding='Latin-1')


In [2]:
autos


dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [3]:
autos.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

Some observations of the dataset:

- Our dataset has 50,000 rows with 20 columns.
- Most of the columns are stored as strings. 
- Some columns that contain dates are also stored as strings.
- A few columns (`vehicleType`, `gearbox`, `model`, `fuelType` and `notRepairedDamage`) have null values, but none have more than 20% null values.
- The column names use [camelcase](https://en.wikipedia.org/wiki/Camel_case) instead of Python's preferred [snakecase](https://en.wikipedia.org/wiki/Snake_case), which means we just cannot replace spaces with underscores.

First, let's check out the first 10 rows of the dataset.


In [4]:
autos.head(10)


dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5  2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6  2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7  2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8  2016-03-22 16:51:34                                         Seat_Arosa   
9  2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   
5  privat   Angebot  $7,900     test         bus                2006   
6  privat   Angebot    $300     test   limousine                1995   
7  privat   Angebot  $1,990  control   limousine                1998   
8  privat   Angebot    $250     test         NaN                2000   
9  privat   Angebot    $590  control         bus                1997   

     gearbox  powerPS    model   odometer  monthOfRegistration fuelType  \
0    manuell      158   andere  150,000km                    3      lpg   
1  automatik      286      7er  150,000km                    6   benzin   
2    manuell      102     golf   70,000km                    7   benzin   
3  automatik       71   fortwo   70,000km                    6   benzin   
4    manuell        0    focus  150,000km                    7   benzin   
5  automatik      150  voyager  150,000km                    4   diesel   
6    manuell       90     golf  150,000km                    8   benzin   
7    manuell       90     golf  150,000km                   12   diesel   
8    manuell        0    arosa  150,000km                   10      NaN   
9    manuell       90   megane  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   
5    chrysler               NaN  2016-03-21 00:00:00             0   
6  volkswagen               NaN  2016-03-20 00:00:00             0   
7  volkswagen              nein  2016-03-16 00:00:00             0   
8        seat              nein  2016-03-22 00:00:00             0   
9     renault              nein  2016-03-16 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50  
5       22962  2016-04-06 09:45:21  
6       31535  2016-03-23 02:48:59  
7       53474  2016-04-07 03:17:32  
8        7426  2016-03-26 18:18:10  
9       15749  2016-04-06 10:46:35

* The columns `dateCrawled`, `dateCreated` and `lastSeen`,   containing dates and time are in a consistent format.
* Some columns (`seller`, `offerType`, `gearBox`, and `notRepairedDamaged`) are in German.
* The `price` and `odometer` columns are numeric values stored as text.
    - The values for `price` are stored with the `$` dollar symbol.
    - The values for `odometer` are stored with the `km` unit.
- Almost all values are the same for `seller`, `offerType`, `notRepairedDamage`, and `nrOfPictures`, and probably will not provide more useful information for further analysis. 

## Cleaning Column Names

We will make two changes here:

- Change the column names from camelcase to snakecase.
- Reword some of the column names based on the data dictionary to be more descriptive.


In [5]:
autos.columns


Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

To convert the column names from camelcase to snakecase, we will use the following function:


In [6]:
def to_snake_case(s):
    snake = "".join(["_"+c.lower() if c.isupper() else c for c in s])
    return snake[1:] if snake.startswith("_") else snake


In [7]:
new_column_names = []

for name in autos.columns:
    new_column_names.append(to_snake_case(name))
    
autos.columns = new_column_names
autos.columns


Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'year_of_registration', 'gearbox', 'power_p_s', 'model',
       'odometer', 'month_of_registration', 'fuel_type', 'brand',
       'not_repaired_damage', 'date_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

The column names are now in Python snakecase naming style. However, we will rename a few more columns to make them even more descriptive.


In [8]:
autos.rename(columns={'abtest': 'ab_test',
                      'year_of_registration': 'registration_year',
                      'month_of_registration': 'registration_month',
                      'not_repaired_damage': 'unrepaired_damage',
                      'date_created': 'ad_created'
                     },
             inplace=True)

autos.columns


Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_p_s', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [9]:
autos.head()


date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_p_s   model   odometer  registration_month fuel_type  \
0    manuell        158  andere  150,000km                   3       lpg   
1  automatik        286     7er  150,000km                   6    benzin   
2    manuell        102    golf   70,000km                   7    benzin   
3  automatik         71  fortwo   70,000km                   6    benzin   
4    manuell          0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

## Initial Data Exploration and Cleaning

Let's do some basic data exploration to determine what other cleaning tasks need to be done. Initially we will look for:

* Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.

* Examples of numeric data stored as text which can be cleaned and converted.

* Categorical data that uses German words that can be translated and mapped to their English counterparts.


In [10]:
autos.describe(include='all')


date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-25 19:57:10  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox     power_p_s  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

Those statistics tell us that.

* The `seller` and `offer_Type` have one value only, which we have observed earlier. These columns will be dropped.

* The `nr_of_pictures` column looks odd, we'll need to investigate this further.

* The `price` and `odometer` columns contain numeric data stored as text (We noticed earlier too!). We will clean and convert them to a numeric data type.

* Based on the minimum and maximum values, `registration_year`, `registration_month` and `power_ps` have obvious outliers. Their values will be corrected or dropped.

* For the `price` column, 1421 out of 50,000 entries have a price of `$0`, that means about `3%` of the vehicles in our dataset are actually free of charge.

* The `unrepaired_damage` column has nearly `20%` of null values. And for the remaining vehicles, about `12%` have unrepaired damages.

Let's investigate the `seller`, `offer_Type` and `nr_of_pictures` columns first.


In [11]:
autos['nr_of_pictures'].value_counts()


0    50000
Name: nr_of_pictures, dtype: int64

In [12]:
autos['seller'].value_counts()


privat        49999
gewerblich        1
Name: seller, dtype: int64

In [13]:
autos['offer_type'].value_counts()


Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

- As noted previously, all (or nearly all) of the values for `seller` and `offerType` are the same, we will drop these columns.

- All the data in `nr_of_pictures` is `0`. It appears that none of the listings include any pictures. We will drop this column along with `seller` and `offerType` now. 


In [14]:
autos.drop(['seller', 'offer_type', 'nr_of_pictures'], 
           axis = 1, 
           inplace = True)


## Translating Non-English Data

Next, we are going to identify categorical data that uses German words, translate them and map the values to their English counterparts. 

After deleting the meaningless columns, we are left with four columns that contain German data: `vehicle_type`, `gear_box`,`fuel_type` and `unrepaired_damage`.  Let's take a look at their vocabulary.


In [15]:
german_columns = ['vehicle_type', 'gearbox','fuel_type', 'unrepaired_damage']

for col in german_columns:
    print(col, ":")
    print(autos[col].unique())
    print('\n')
    

vehicle_type :
['bus' 'limousine' 'kleinwagen' 'kombi' nan 'coupe' 'suv' 'cabrio'
 'andere']


gearbox :
['manuell' 'automatik' nan]


fuel_type :
['lpg' 'benzin' 'diesel' nan 'cng' 'hybrid' 'elektro' 'andere']


unrepaired_damage :
['nein' nan 'ja']




Let's create a `German dictionary` for these words and map them into their English equivalent.


In [16]:
german_dict = {
    ## vehicle_type 
    'bus': 'bus',
    'limousine': 'limousine',
    'kleinwagen': 'small car', 
    'kombi': 'station wagon',
    'coupe': 'coupe',
    'suv': 'suv',
    'cabrio': 'convertible',
    'andere': 'other',
    ## gearbox
    'manuell': 'manual',
    'automatik': 'automatic',
    # fuel_type
    'lpg': 'lpg',
    'benzin': 'petrol',
    'diesel': 'diesel',
    'cng': 'cng',
    'hybrid': 'hybrid',
    'elektro': 'electric',
    ## unrepaired_damage :
    'nein': 'no',
    'ja': 'yes'
}

### Map German words into English
for column in german_columns:
    
    print('Before translation.', '\n')
    print(autos[column].value_counts(dropna=False))
    
    print('-'*30)
    print('Translating...', '\n')
    autos[column] = autos[column].map(german_dict)
    
    print('-'*30)
    print('After translation.', '\n')
    print(autos[column].value_counts(dropna=False))
    print('-'*30)
    print('-'*30)
    print('\n')
    

Before translation. 

limousine     12859
kleinwagen    10822
kombi          9127
NaN            5095
bus            4093
cabrio         3061
coupe          2537
suv            1986
andere          420
Name: vehicle_type, dtype: int64
------------------------------
Translating... 

------------------------------
After translation. 

limousine        12859
small car        10822
station wagon     9127
NaN               5095
bus               4093
convertible       3061
coupe             2537
suv               1986
other              420
Name: vehicle_type, dtype: int64
------------------------------
------------------------------


Before translation. 

manuell      36993
automatik    10327
NaN           2680
Name: gearbox, dtype: int64
------------------------------
Translating... 

------------------------------
After translation. 

manual       36993
automatic    10327
NaN           2680
Name: gearbox, dtype: int64
------------------------------
------------------------------


Befor

The German words have been translated to their English equivalent, and their value counts remain the same. We can conclude that the translation process was successfully carried out.

## Exploring the `Odometer` and `Price` Columns

We have already noticed that the "price" and "odometer" columns are numeric values stored as text. For these columns, we will remove any non-numeric characters, convert the columns to a numeric data type, and rename the columns to preserve their unit information.


In [17]:
autos['price'] = (autos['price']
                      .str.replace('$','')
                      .str.replace(',','')
                      .astype(int)
                 )

autos.rename({'price': 'price_dollars'}, axis=1, inplace=True)

autos['price_dollars'].head()


0    5000
1    8500
2    8990
3    4350
4    1350
Name: price_dollars, dtype: int64

In [18]:
autos['odometer'] = (autos['odometer']
                      .str.replace('km','')
                      .str.replace(',','')
                      .astype(int)
                 )

autos.rename({'odometer': 'odometer_km'}, axis=1, inplace=True)

autos['odometer_km'].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

Let's continue exploring the data for outliers.  We will start by analysing the "odometer_km" and "price_usd" numeric columns, using the minimum and maximum values that appear too high or too low to be realistic and that we may want to remove.


In [19]:
autos['price_dollars'].unique().shape

(2357,)

In [20]:
autos['odometer_km'].unique().shape

(13,)

The `price_usd` column has 2357 unique values, and 13 for the `odometer_km` column.

Let's take a look at `odometer_km` statistics.


In [21]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [22]:
autos['odometer_km'].value_counts().sort_index()

5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64

- We can see that the values in this field are rounded, which might indicate that sellers had to choose from pre-set options for this field. 
- There are no outliers in this column since the values in this column are between 5,000 and 150,000 km. 
- The average mileage is 125,732km. 
- Approximately 65 per cent of vehicles have significant mileage. This means, there is more high mileage than low mileage vehicles on sale.

So let's check out `price_usd`.


In [23]:
autos['price_dollars'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_dollars, dtype: float64

In [24]:
autos['price_dollars'].value_counts().head(20)

0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
700      395
4500     394
300      384
2200     382
950      379
Name: price_dollars, dtype: int64

- Again, the prices in this column seem rounded. Given there are 2357 unique values in the column, that may just be people's tendency to round prices on this site.
- There are 1421 vehicles listed with `$0` price. We might consider removing these rows, considering this is only `2%` of the data set.
- The maximum price is one hundred million dollars, which seems a lot for a used vehicle. 

Let's look at the prices further.


In [25]:
autos['price_dollars'].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price_dollars, dtype: int64

In [26]:
autos['price_dollars'].value_counts().sort_index(ascending=False).tail(20)

35       1
30       7
29       1
25       5
20       4
18       1
17       3
15       2
14       1
13       2
12       3
11       2
10       7
9        1
8        1
5        2
3        1
2        3
1      156
0     1421
Name: price_dollars, dtype: int64

There are a number of listings with prices below `$30`, including 1,421 at `$0`. There are also a small number of listings with unrelistic high values, including 14 at around `$1 million`.

Since eBay is an aution site, there may legitimately be items in which the opening bid is `$1`. As a result, we will keep the `$1` items, but we will remove anything over `$350,000`, as it appears that prices rise steadily to that number and then rise to a less realistic number.


In [27]:
# Removing the outlier
autos = autos[autos["price_dollars"].between(1,351000)]
autos["price_dollars"].describe()


count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price_dollars, dtype: float64

## Exploring the Date Columns

Now let's turn to the date columns and understand the date range of the data.

There are 5 columns with date information:

- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

The first three of these columns are identified as string values by pandas, so we need to convert them into a numerical representation for further analysis. The other two columns are already represented as numeric values. 

Let's begin by understanding how the values in the three string columns are formatted.


In [28]:
autos[['date_crawled','ad_created','last_seen']].head()


date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

These columns represent full timestamp value, with the first 10 characters represent the day in the yy-mm-dd format (e.g. 2016-03-26). The rest of the characters represent the time. We are interested in the `day` part only.
 
We are going to calculate the distribution of values in the `date_crawled`, `ad_created`, and `last_seen` columns as percentages, including missing values and ranking by date from earliest to latest. 


In [29]:
# Counts the crawl dates
print(autos['date_crawled'].str[:10].unique().size) 

date_crawled = (autos['date_crawled']
         .str[:10]
         .value_counts(normalize=True, dropna=False)
         .sort_index()
        )

date_crawled

34


2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

In [30]:
'''

import matplotlib.pyplot as plt

dates = list(date_crawled.index)
percent_crawled = list(date_crawled.values)

plt.plot(dates,percent_crawled) 
plt.title('Ad Crawled Per Day (March - April 2016)')
plt.xlabel('Dates')
plt.ylabel('Ad Crawled (%)')
plt.show()
'''

"\n\nimport matplotlib.pyplot as plt\n\ndates = list(date_crawled.index)\npercent_crawled = list(date_crawled.values)\n\nplt.plot(dates,percent_crawled) \nplt.title('Ad Crawled Per Day (March - April 2016)')\nplt.xlabel('Dates')\nplt.ylabel('Ad Crawled (%)')\nplt.show()\n"

- It appears the crawler collected data from the site everyday from the 5th of March to the 7th of April 2016, a period of 34 days. 

- The number of listings crawled on each day seems quite consistent. 


In [31]:
# Counts the ad created dates
print('Ad created dates:', autos['ad_created'].str[:10].unique().size) 

ad_created = (autos['ad_created']
         .str[:10]
         .value_counts(normalize=True, dropna=False)
         .sort_index()
        )


Ad created dates: 76


In [32]:
print(ad_created[:30])


2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
Name: ad_created, dtype: float64


In [33]:
print(ad_created[30:])

2016-02-22    0.000021
2016-02-23    0.000082
2016-02-24    0.000041
2016-02-25    0.000062
2016-02-26    0.000041
2016-02-27    0.000124
2016-02-28    0.000206
2016-02-29    0.000165
2016-03-01    0.000103
2016-03-02    0.000103
2016-03-03    0.000865
2016-03-04    0.001483
2016-03-05    0.022897
2016-03-06    0.015320
2016-03-07    0.034737
2016-03-08    0.033316
2016-03-09    0.033151
2016-03-10    0.031895
2016-03-11    0.032904
2016-03-12    0.036755
2016-03-13    0.017008
2016-03-14    0.035190
2016-03-15    0.034016
2016-03-16    0.030125
2016-03-17    0.031278
2016-03-18    0.013590
2016-03-19    0.033687
2016-03-20    0.037949
2016-03-21    0.037579
2016-03-22    0.032801
2016-03-23    0.032060
2016-03-24    0.029280
2016-03-25    0.031751
2016-03-26    0.032266
2016-03-27    0.030989
2016-03-28    0.034984
2016-03-29    0.034037
2016-03-30    0.033501
2016-03-31    0.031875
2016-04-01    0.033687
2016-04-02    0.035149
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05 

- From our data set, the first listing was created on 11th of June 2015, and the last on 7th of April 2016.

- The distribution of values is very low at the beginning and then grow steadily from the 5th of March 2016, when the ads were first crawled from the website. 


In [34]:
# Counts the last seen dates
print('Last seen dates:', autos['last_seen'].str[:10].unique().size) 

last_seen = (autos['last_seen']
         .str[:10]
         .value_counts(normalize=True, dropna=False)
         .sort_index()
        )

last_seen

Last seen dates: 34


2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

- The date range when the crawler last saw the ads is also from March 5, 2016 to April 7, 2016, as for date_crawled.

- The distribution of values is basically constant, increasing in the last three days only.

Now we'll take a look at registration_year.

In [35]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64



One thing that stands out from the `registration_year` statistics is that there are some odd values:

- The minimum value is `1000`, before cars were invented.

- The maximum value is `9999`, many years into the future.

## Dealing with Incorrect Registration Year Data

Because a car cannot be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s. 

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic. 


In [36]:
autos['registration_year'].value_counts().sort_index()


1000    1
1001    1
1111    1
1800    2
1910    5
       ..
5911    1
6200    1
8888    1
9000    1
9999    3
Name: registration_year, Length: 95, dtype: int64

It seems the earliest plausible registration year is 1910. Let's investigate further for these vehicles.


In [37]:
autos[autos['registration_year'] <= 1910]

date_crawled                                   name  \
3679   2016-04-04 00:36:17                             Suche_Auto   
10556  2016-04-01 06:02:10                             UNFAL_Auto   
22316  2016-03-29 16:56:41  VW_Kaefer.__Zwei_zum_Preis_von_einem.   
22659  2016-03-14 08:51:18                           Opel_Corsa_B   
24511  2016-03-17 19:45:11            Trabant__wartburg__Ostalgie   
28693  2016-03-22 17:48:41                         Renault_Twingo   
30781  2016-03-25 13:47:46       Opel_Calibra_V6_DTM_Bausatz_1:24   
32585  2016-04-02 16:56:39                             UNFAL_Auto   
45157  2016-03-11 22:37:01                             Motorhaube   
49283  2016-03-15 18:38:53                             Citroen_HY   

       price_dollars  ab_test vehicle_type  registration_year gearbox  \
3679               1     test          NaN               1910     NaN   
10556            450  control          NaN               1800     NaN   
22316           1500  control          NaN               1000  manual   
22659            500     test          NaN               1910     NaN   
24511            490  control          NaN               1111     NaN   
28693            599  control    small car               1910  manual   
30781             30     test          NaN               1910     NaN   
32585            450  control          NaN               1800     NaN   
45157             15  control          NaN               1910     NaN   
49283           7750  control          NaN               1001     NaN   

       power_p_s    model  odometer_km  registration_month fuel_type  \
3679           0      NaN         5000                   0       NaN   
10556       1800      NaN         5000                   2       NaN   
22316          0   kaefer         5000                   0    petrol   
22659          0    corsa       150000                   0       NaN   
24511          0      NaN         5000                   0       NaN   
28693          0      NaN         5000                   0    petrol   
30781          0  calibra       100000                   0       NaN   
32585       1800      NaN         5000                   2       NaN   
45157          0      NaN         5000                   0       NaN   
49283          0   andere         5000                   0       NaN   

                brand unrepaired_damage           ad_created  postal_code  \
3679   sonstige_autos               NaN  2016-04-04 00:00:00        40239   
10556      mitsubishi                no  2016-04-01 00:00:00        63322   
22316      volkswagen               NaN  2016-03-29 00:00:00        48324   
22659            opel               NaN  2016-03-14 00:00:00        52393   
24511         trabant               NaN  2016-03-17 00:00:00        16818   
28693         renault               NaN  2016-03-22 00:00:00        70376   
30781            opel               NaN  2016-03-25 00:00:00        47638   
32585      mitsubishi                no  2016-04-02 00:00:00        63322   
45157         trabant               NaN  2016-03-11 00:00:00        90491   
49283         citroen               NaN  2016-03-15 00:00:00        66706   

                 last_seen  
3679   2016-04-04 07:49:15  
10556  2016-04-01 09:42:30  
22316  2016-03-31 10:15:28  
22659  2016-04-03 07:53:55  
24511  2016-04-07 07:17:29  
28693  2016-04-06 09:16:59  
30781  2016-03-26 23:46:29  
32585  2016-04-04 14:46:21  
45157  2016-03-25 11:18:57  
49283  2016-04-06 18:47:20

- Certain columns in these entries have missing or erroneous values.

- `Odometer_km` seems unrealistically low for vehicles over 100 years old.

- The entry `22316` was registered on `1000` and was made by `Volkswagen`, a German company founded in `1937`.
    
- Prices are also low, around `500` for seven of these `veteran cars`.
    
These incorrect values suggest these entries do not belong to the data set and can be removed.

Let's investigate the next three entries from our exploration: 1927, 1929, and 1931.  


In [38]:
autos[autos['registration_year'] == 1927]

date_crawled                     name  price_dollars  ab_test  \
21416  2016-03-12 08:36:21  Essex_super_six__Ford_A          16500  control   

      vehicle_type  registration_year gearbox  power_p_s   model  odometer_km  \
21416  convertible               1927  manual         40  andere         5000   

       registration_month fuel_type brand unrepaired_damage  \
21416                   5    petrol  ford               NaN   

                ad_created  postal_code            last_seen  
21416  2016-03-12 00:00:00        74821  2016-03-15 12:45:12

In [39]:
autos[autos['registration_year'] == 1929]

date_crawled        name  price_dollars ab_test vehicle_type  \
22101  2016-03-09 16:51:17  BMW_Andere          11500    test  convertible   

       registration_year gearbox  power_p_s   model  odometer_km  \
22101               1929  manual         15  andere         5000   

       registration_month fuel_type brand unrepaired_damage  \
22101                   1       NaN   bmw               yes   

                ad_created  postal_code            last_seen  
22101  2016-03-09 00:00:00        70569  2016-04-07 06:17:11

In [40]:
autos[autos['registration_year'] == 1931]

date_crawled                               name  price_dollars  \
11246  2016-03-26 19:49:59  Ford_Model_A_Roadster_Deluxe_1931          27500   

       ab_test vehicle_type  registration_year gearbox  power_p_s   model  \
11246  control  convertible               1931  manual         39  andere   

       odometer_km  registration_month fuel_type brand unrepaired_damage  \
11246        10000                   7    petrol  ford                no   

                ad_created  postal_code            last_seen  
11246  2016-03-26 00:00:00         9322  2016-04-06 09:46:59

The values for these three entries all seem plausible: high prices, convertible vehicle type, manual gearbox, low power and low odometer values.

It seems reasonable, then, to use the year 1927 as the first year and 2016 as the final year. 

Let's determine what percentage of our data has invalid values in this column:

In [41]:
# Check the percentage of entries between 1927 and 2016
(~autos['registration_year'].between(1927,2016)).sum() / autos.shape[0] * 100


3.8896324513538554

Given that this is less than 4% of our data, we will remove these entries.


In [42]:
autos = autos[autos['registration_year'].between(1927, 2016)]


In [43]:
autos['registration_year'].value_counts(normalize=True).sort_index()


1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000043
1937    0.000086
          ...   
2012    0.028066
2013    0.017204
2014    0.014204
2015    0.008398
2016    0.026138
Name: registration_year, Length: 77, dtype: float64

In [44]:
autos['registration_year'].describe()


count    46676.000000
mean      2002.920709
std          7.120843
min       1927.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

It appears that most of the vehicles were first registered in the past 20 years.

## Exploring Price by Brand

Since we are working with data on cars, it is natural to explore the price variations across different car brands. To start, let take a look at the unique car brands available in our dataset. 


In [45]:
brands = autos['brand'].value_counts(normalize=True)
brands 


volkswagen        0.211286
bmw               0.110057
opel              0.107550
mercedes_benz     0.096474
audi              0.086576
ford              0.069907
renault           0.047133
peugeot           0.029844
fiat              0.025645
seat              0.018275
skoda             0.016411
nissan            0.015276
mazda             0.015190
smart             0.014161
citroen           0.014011
toyota            0.012705
hyundai           0.010027
sonstige_autos    0.009791
volvo             0.009148
mini              0.008763
mitsubishi        0.008227
honda             0.007841
kia               0.007070
alfa_romeo        0.006642
porsche           0.006127
suzuki            0.005935
chevrolet         0.005699
chrysler          0.003514
dacia             0.002635
daihatsu          0.002507
jeep              0.002271
subaru            0.002142
land_rover        0.002100
saab              0.001650
jaguar            0.001564
daewoo            0.001500
trabant           0.001371
r

German manufacturers represent four out of the top five brands, almost 50% of the total listings. Volkwagen is by far the most popular brand, with approximately double the cars for sale of the next two brands (BMW and Opel) combined.

There are lots of brands that don't have a significant percentage of listings, so we will limit our analysis to brands representing more than 5% of total listings.


In [46]:
# Finding brands with 5% or more of the total number of vehicles
top_brands = brands[brands > 0.05].index
top_brands 


Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')

In [47]:
# Calculate the average price for these brands

brand_mean_prices = {}

for brand in top_brands:
    selected_brand = autos[autos['brand'] == brand]
    mean_price = selected_brand['price_dollars'].mean()
    
    brand_mean_prices[brand] = int(mean_price)
    
brand_mean_prices


{'volkswagen': 5402,
 'bmw': 8332,
 'opel': 2976,
 'mercedes_benz': 8628,
 'audi': 9336,
 'ford': 3749}

Of the top six brands, there is a distinct price gap:

- Audi, BMW and Mercedes Benz are more expensive

- Ford and Opel are less expesive

- Volkwagen, which seems to be in the middle of these prices, is not a surprise, it is the most popular brand in our data set.


## Exploring Mileage 

For the top six brands, let's use aggregation to understand the average mileage for those cars and if there is any visible link with mean price. 


In [48]:
# Calculate the average mileage for these six brands
brand_mean_mileage = {}

for brand in top_brands:
    brand_only = autos[autos['brand'] == brand]
    mean_mileage = brand_only['odometer_km'].mean()
    
    brand_mean_mileage[brand] = int(mean_mileage)


In [49]:
# Convert the average mileage dictionary into a Pandas Series object
bmm_series = pd.Series(brand_mean_mileage).sort_values(ascending=False)
print(bmm_series)


bmw              132572
mercedes_benz    130788
opel             129311
audi             129157
volkswagen       128707
ford             124266
dtype: int64


In [50]:
# Convert the average price dictionary into a Pandas Series object
bmp_series = pd.Series(brand_mean_prices)
print(bmp_series)

volkswagen       5402
bmw              8332
opel             2976
mercedes_benz    8628
audi             9336
ford             3749
dtype: int64


In [51]:
# Convert the mean prices Series object to a DataFrame
df = pd.DataFrame(bmp_series, columns=['mean_price'])

# Combine both Series objects into a single dataframe
df['mean_mileage'] = bmm_series
df


mean_price  mean_mileage
volkswagen           5402        128707
bmw                  8332        132572
opel                 2976        129311
mercedes_benz        8628        130788
audi                 9336        129157
ford                 3749        124266

Our expectation is that car prices will go down as mileage increases. However, we find that is not the case. The more expensive brands (Audi, BMW, and Mercedes Benz) have higher average mileages, while the less expensive brands (Ford, Opel and Volkswagen) have slightly lower mileages. 

Moreover, the mileage range does not vary as much as the prices of these brands, with approximately 7% range variation.

## Finding the Most Common Brand/Model Combinations

We already know which brands are the most common in our data set. Let's find out which specific models contribute to the popularity of the brand.


In [52]:
brand_model = autos[['brand','model']]

brand_model_grouped = brand_model.groupby('brand')['model'].value_counts()
brand_model_grouped


brand       model  
alfa_romeo  156        88
            147        80
            andere     60
            159        32
            spider     32
                       ..
volvo       v50        29
            850        28
            c_reihe    28
            s60        17
            v60         3
Name: model, Length: 290, dtype: int64

It is not easy to read our results. By default, the .groupby() method returns a series indexed by multiple columns. We can improve the readability of our results by resetting the indexes of the grouped columns back into dataframe columns. Our aggregate column can also be given an appropriate name:

In [53]:
# Reset multi-index series to dataframe and assign a name to the aggregate column
brand_model_grouped = brand_model_grouped.reset_index(name='count') 
brand_model_grouped.head(5)


brand   model  count
0  alfa_romeo     156     88
1  alfa_romeo     147     80
2  alfa_romeo  andere     60
3  alfa_romeo     159     32
4  alfa_romeo  spider     32

We only need to know the most popular cars, so we will sort the entire dataset in descending order based on the `counts` column:

In [54]:
brand_model_grouped.sort_values(by='count', axis=0, ascending=False).head(15)


brand        model  count
259     volkswagen         golf   3707
22             bmw          3er   2615
260     volkswagen         polo   1609
176           opel        corsa   1591
261     volkswagen       passat   1349
177           opel        astra   1348
6             audi           a4   1231
138  mercedes_benz     c_klasse   1136
23             bmw          5er   1132
139  mercedes_benz     e_klasse    958
7             audi           a3    825
8             audi           a6    797
77            ford        focus    762
78            ford       fiesta    722
262     volkswagen  transporter    674

The following are the most popular models within each of the brands:

- Golf, Polo, and Passat are the most popular Volkswagen models.

- Astra and Corsa are the most popular Opel models.

- Audi A-series (A3, A4 and A6) are extremely popular.

- Mercedes' C and E class cars are popular.

- 3 and 5 series BMW are most popular.

- Focus and fiesta are Ford's most popular models.


## Analyzing Price based on Mileage

We will now look into how mileage affects car prices. We will break our odometer readings into intervals and look for any noteworthy price patterns. Let's look at the `odometer_km` values that are unique:


In [55]:
sorted(autos['odometer_km'].unique())


[5000,
 10000,
 20000,
 30000,
 40000,
 50000,
 60000,
 70000,
 80000,
 90000,
 100000,
 125000,
 150000]

This column has a small number of unique values, and the data appears to be rounded to the nearest thousands. Our odometer readings can be divided into seven intervals as follows:

- 5000km - 10000km 
- 20000km - 30000km
- 40000km - 50000km
- 60000km - 70000km
- 80000km - 90000km
- 100000km - 125000km
- 150000km or more


In [56]:
# extract relevant columns
mileage = autos[['odometer_km', 'price_dollars']].copy()

odometer_groups = {
    5000:'5000km - 10000km',
    10000:'5000km - 10000km',
    20000:'20000km - 30000km',
    30000:'20000km - 30000km',
    40000:'40000km - 50000km',
    50000:'40000km - 50000km',
    60000:'60000km - 70000km',
    70000:'60000km - 70000km',
    80000:'80000km - 90000km',
    90000:'80000km - 90000km',
    100000:'100000km - 125000km',
    125000:'100000km - 125000km',
    150000:'150000km or more'
}

mileage['odometer_km'] = mileage['odometer_km'].map(odometer_groups)

price_by_mileage = (mileage.groupby('odometer_km')['price_dollars'].mean()
                       .reset_index(name='avg_price_dollars')
                       .sort_values(by='avg_price_dollars', axis=0, ascending=False)
                  )

price_by_mileage


odometer_km  avg_price_dollars
2    20000km - 30000km       17517.633822
3    40000km - 50000km       14563.488268
4     5000km - 10000km       11649.906158
5    60000km - 70000km       11637.516631
6    80000km - 90000km        9032.042323
0  100000km - 125000km        6786.023431
1     150000km or more        3768.035733

With one exception, we can see that average prices follow a predictable trend.

- The average sell prices for `5000 km to 10000 km` and `60000 km to 70000 km` cars are approximately the same, at around `$11640`. Lower mileage vehicles are expected to sell for more money. Would a car with a mileage of less than 10,000 km be an outlier? Or was it because some vehicles were damaged immediately after purchase?

- As expected, cars with higher mileage `150,000 km or more` are the cheapest.

- Cars with a lower average mileage are the most expensive. This category includes cars with a mileage of `20,000 to 50,000` kilometres.

- Cars with moderate to high mileage `60,000 km to 120,000 km` are next, with an average price of `$9151`.


## Comparing Prices of Damaged vs. Non-Damaged Cars

Our final step will be to determine how much cars with damage cost less than cars without damage. Let's start by counting how many vehicles have unrepaired damages.


In [57]:
print('Counts')
print('-'*30)
print(autos['unrepaired_damage'].value_counts(dropna=False))
print('')
print('Percentage')
print('-'*30)
print(autos['unrepaired_damage'].value_counts(dropna=False, normalize=True))


Counts
------------------------------
no     33834
NaN     8302
yes     4540
Name: unrepaired_damage, dtype: int64

Percentage
------------------------------
no     0.724869
NaN    0.177864
yes    0.097266
Name: unrepaired_damage, dtype: float64


Almost 10% of the cars in our data set have unrepaired defects, whereas the majority of the cars (around 72%) are in good condition. We don't know what happened to the remaining 18%. We'll now compute the average price for both categories.


In [58]:
autos.groupby('unrepaired_damage')['price_dollars'].mean()


unrepaired_damage
no     7164.033103
yes    2241.146035
Name: price_dollars, dtype: float64

The price difference is significant: damaged vehicles cost 3.5 times less than non-damaged cars on average. Let's take a look at the most costly car among the damaged ones just for fun.


In [59]:
# Find the most expensive car among the damaged ones
autos.loc[autos[autos['unrepaired_damage'] == 'yes']['price_dollars'].idxmax()]


date_crawled                        2016-03-21 18:36:18
name                  Vw_troureg_4_2__V_8_Diesel_Unfall
price_dollars                                     44200
ab_test                                            test
vehicle_type                                        suv
registration_year                                  2013
gearbox                                       automatic
power_p_s                                           370
model                                           touareg
odometer_km                                       60000
registration_month                                    8
fuel_type                                        diesel
brand                                        volkswagen
unrepaired_damage                                   yes
ad_created                          2016-03-21 00:00:00
postal_code                                       50997
last_seen                           2016-03-22 23:18:30
Name: 38870, dtype: object

It's puzzling why this Volkswagen Touareg, which is three years old, is advertised at `$44200` in 2016. It has unrepaired faults and 60,000 kilometres on the odometer. Despite this, it is priced similarly to a new vehicle. We will not devote any more time to this car at this time.

## Conclusion

To conclude, we cleaned and investigated a dataset of used vehicles from the German eBay website in this research. Renaming columns and giving them descriptive names according to Python tradition, deleting non-informative columns, mapping German words to English, converting strings to numeric data types, removing outliers, and fixing year range were among the cleaning activities. 

We discovered that the adverts were uploaded to the website from June 2015 to April 2016, and that they were crawled from March 2016 to April 2016. 

The most common brands in our sample were determined, with Volkswagen taking first place with a significant lead over the others. 

Only six brands account for more than 5\% of total vehicle sales. Average costs and mileage values were calculated for these top brands. The most common brand / model combination is Volkswagen Golf, and 65 percent of cars have the higher mileage (150,000 km). The price patterns depending on mileage were examined. 

We discovered that cars with an average mileage of 40,000–50,000 km have the highest average pricing. 

Finally, we looked at the average price difference between non-damaged and damaged autos and discovered that the latter are 3.5 times cheaper.
